lib/datadog/statsd/message_buffer.rb

```diff
@@ -28,30 +39,53 @@ def add(message)
         return nil unless message_size > 0 # to avoid adding empty messages to the buffer
         return nil unless ensure_sendable!(message_size)
 
-        flush if should_flush?(message_size)
+        @mon.synchronize {
```

Comment:
`add` is in the hot-path of the library when running at high-throughput, and, while I agree we need some synchronization here (i.e. [I have a draft with a simpler multi-threaded Sender code, and it needs a `Mutex` in `add` as well](https://github.com/DataDog/dogstatsd-ruby/pull/209/files#diff-22ef838ef4e2984fdedbeae928e214979fa440026ddc2f3f96b90328098b5ee8R61)), I'm wondering if using a `Monitor` is the best thing to do here, as:
  * it can be slightly more confusing than a mutex since Monitor are reentrant ([see this post](https://japgolly.blogspot.com/2012/04/ruby-mutex-reentrancy.html))
  * could be slower than a mutex

Reply:
I think it is better to use `Monitor` for the following reasons:

* It has better performance than `Mutex`
    - As described on [the release note of Ruby 2.7.0](https://www.ruby-lang.org/en/news/2019/12/25/ruby-2-7-0-released/), the performance of `Monitor` has been improved. (cf. https://github.com/ruby/ruby/commit/caac5f777ae288b5982708b8690e712e1cae0cf6)
* It makes code clear
    - We don't have to manage who is the owner of the lock
* It is also used in [Fluentd](https://github.com/fluent/fluentd), which also processes so many events

Actually, the following benchmark shows `Monitor` has better performance than `Mutex`. Of course, the lock makes the performance worse than before, but I think `MessageBuffer` still has good performance.

```ruby
require 'monitor'
require 'benchmark/ips'

class MessageBufferBase
  PAYLOAD_SIZE_TOLERANCE = 0.05

  def initialize
    @max_payload_size = 1_432
    @max_pool_size = Float::INFINITY
    @overflowing_stategy = :drop

    @buffer = String.new
    clear_buffer

    @closed = false
  end

  private

  attr :max_payload_size
  attr :max_pool_size

  attr :overflowing_stategy
  attr_reader :flush_interval

  attr :buffer

  def _flush
    return if buffer.empty?

    clear_buffer
  end

  def should_flush?(message_size)
    return true if buffer.bytesize + 1 + message_size >= max_payload_size

    false
  end

  def clear_buffer
    buffer.clear
    @message_count = 0
  end

  def preemptive_flush?
    @message_count == max_pool_size || buffer.bytesize > bytesize_threshold
  end

  def ensure_sendable!(message_size)
    return true if message_size <= max_payload_size

    if overflowing_stategy == :raise
      raise Error, 'Message too big for payload limit'
    end

    false
  end

  def bytesize_threshold
    @bytesize_threshold ||= (max_payload_size - PAYLOAD_SIZE_TOLERANCE * max_payload_size).to_i
  end
end

class MessageBufferWithoutLock < MessageBufferBase
  def add(message)
    message_size = message.bytesize

    return nil unless message_size > 0 # to avoid adding empty messages to the buffer
    return nil unless ensure_sendable!(message_size)

    raise Error, 'buffer is closed' if @closed
    flush if should_flush?(message_size)

    buffer << "\n" unless buffer.empty?
    buffer << message

    @message_count += 1

    # flush when we're pretty sure that we won't be able
    # to add another message to the buffer
    flush if preemptive_flush?

    true
  end

  def flush
    _flush
  end
end

class MessageBufferWithMutex < MessageBufferBase
  def initialize
    super
    @mutex = Mutex.new
    @mutex_owner = nil
  end

  def add(message)
    message_size = message.bytesize

    return nil unless message_size > 0 # to avoid adding empty messages to the buffer
    return nil unless ensure_sendable!(message_size)

    @mutex.synchronize {
      begin
        @mutex_owner = Thread.current
        raise Error, 'buffer is closed' if @closed
        flush if should_flush?(message_size)

        buffer << "\n" unless buffer.empty?
        buffer << message

        @message_count += 1

        # flush when we're pretty sure that we won't be able
        # to add another message to the buffer
        flush if preemptive_flush?
      ensure
        @mutex_owner = nil
      end
    }

    true
  end

  def flush
    if @mutex_owner != Thread.current
      @mutex.lock
      locked = true
    end
    _flush
  ensure
    @mutex.unlock if locked
  end
end

class MessageBufferWithMonitor < MessageBufferBase
  def initialize
    super
    @mon = Monitor.new
  end

  def add(message)
    message_size = message.bytesize

    return nil unless message_size > 0 # to avoid adding empty messages to the buffer
    return nil unless ensure_sendable!(message_size)

    @mon.synchronize {
      raise Error, 'buffer is closed' if @closed
      flush if should_flush?(message_size)

      buffer << "\n" unless buffer.empty?
      buffer << message

      @message_count += 1

      # flush when we're pretty sure that we won't be able
      # to add another message to the buffer
      flush if preemptive_flush?
    }

    true
  end

  def flush
    @mon.synchronize { _flush }
  end
end


Benchmark.ips do |x|
  b1 = MessageBufferWithoutLock.new
  b2 = MessageBufferWithMutex.new
  b3 = MessageBufferWithMonitor.new

  messages = [
    "a" * 20,
    "a" * 30,
    "a" * 40,
  ]

  x.report('MessageBufferWithoutLock#add') { messages.each { |m| b1.add(m) } }
  x.report('MessageBufferWithMutex#add') { messages.each { |m| b2.add(m) } }
  x.report('MessageBufferWithMonitor#add') { messages.each { |m| b3.add(m) } }
  x.compare!
end
```

The result in my environment (macOS 11.6.1, 2.3 GHz Quad-Core Intel Core i7, ruby 3.0.2p107, ) is here:

```
Warming up --------------------------------------
MessageBufferWithoutLock#add
                        57.506k i/100ms
MessageBufferWithMutex#add
                        42.935k i/100ms
MessageBufferWithMonitor#add
                        38.971k i/100ms
Calculating -------------------------------------
MessageBufferWithoutLock#add
                        566.378k (± 2.4%) i/s -      2.875M in   5.079683s
MessageBufferWithMutex#add
                        417.459k (± 7.1%) i/s -      2.104M in   5.080333s
MessageBufferWithMonitor#add
                        415.527k (± 3.5%) i/s -      2.104M in   5.070831s

Comparison:
MessageBufferWithoutLock#add:   566377.6 i/s
MessageBufferWithMutex#add:   417458.5 i/s - 1.36x  (± 0.00) slower
MessageBufferWithMonitor#add:   415527.0 i/s - 1.36x  (± 0.00) slower

```

Reply:
In the benchmark, I don't consider the flush thread, because I assume that it calls `flush` much less frequently than `add`, which is called by the sender thread.

Reply:
If you prefer Mutex, I'll change the code in the same way as the benchmark.

Reply:
Thanks for the benchs @abicky and having took the time to share your thoughts, that's very convincing 💯 

I'll do a second complete review ASAP @djmitche with process forking in mind while reviewing (forks are heavily occurring for people using Puma, Sidekiq, etc...).

Reply:
Are we sure that these locks are per process and not system-wide with a monitor? I guess so but just to make sure.
Since workers libraries rely on forks, this lock can't be system-wide for our usage.

Reply:
If I understand @remeh, the issue you're concerned with is that the main thread calls `fork()` while another thread is in the midst of executing an `add` call and is holding the lock.  In the child process, that thread would not exist but the lock would still be held, so subsequent calls to `add` would hang.

Reply:
I'm sorry for my late reply. 

> Are we sure that these locks are per process and not system-wide with a monitor?

Yes. As you can see below, the monitor in the child process is unlocked:

```ruby
require 'monitor'

mon = Monitor.new
thread = Thread.new do
  loop do
    puts "pid: #{Process.pid}, mon.mon_locked?: #{mon.mon_locked?}"
    sleep(1)
  end
end
mon.mon_enter

child_pid = fork do
  puts "child process. pid: #{Process.pid}, thread.alive?: #{thread.alive?}, mon.mon_locked?: #{mon.mon_locked?}"
  mon.mon_enter
  puts "sleep a little"
  sleep(3)
end

puts "parent process. pid: #{Process.pid}, child pid: #{child_pid}, thread.alive?: #{thread.alive?}, mon.mon_locked?: #{mon.mon_locked?}"

Process.waitpid(child_pid)
```

The output is here:

```
parent process. pid: 46387, child pid: 46417, thread.alive?: true, mon.mon_locked?: true
pid: 46387, mon.mon_locked?: true
child process. pid: 46417, thread.alive?: false, mon.mon_locked?: true
sleep a little
pid: 46387, mon.mon_locked?: true
pid: 46387, mon.mon_locked?: true
```

Reply:
By the way, I've implemented the feature in another way https://github.com/DataDog/dogstatsd-ruby/pull/231. I'd like you to merge one you prefer and close the other.